File: Chapter12/VB/HostEmulator/Workers/ConnectionProcessor.vb
Imports System.Net
Imports System.Net.Sockets

Public Class ConnectionProcessor
  Private _socket As Socket

  Public Sub New(ByVal theSocket As Socket)
    _socket = theSocket
  End Sub

  Public Sub WaitForRequests()
    While True
      If _socket.Available > 0 Then
        FireRequestStarted(_socket)
      End If

      FireIdle()
    End While
  End Sub

  Public Event OnRequestStarted(ByVal theSocket As Socket)
  Public Sub FireRequestStarted(ByVal theSocket As Socket)
    RaiseEvent OnRequestStarted(theSocket)
  End Sub

  Public Event OnRequestProcessed(ByVal theRequest As String, ByVal theSequenceNumber As Integer, ByVal theDuration As Integer, ByVal theResponse As String)
  Public Sub FireRequestProcessed(ByVal theRequest As String, ByVal theSequenceNumber As Integer, ByVal theDuration As Integer, ByVal theResponse As String)
    RaiseEvent OnRequestProcessed(theRequest, theSequenceNumber, theDuration, theResponse)
  End Sub

  Public Event OnIdle()
  Public Sub FireIdle()
    RaiseEvent OnIdle()
  End Sub
End Class

File: Chapter12/C#/HostEmulator/Workers/ConnectionProcessor.cs
using System;
using System.Net.Sockets;

namespace HostEmulator
{
	/// <summary>
	/// ConnectionProcessor waits for incoming requests on a given connection.
	/// When a request arrives, ConnectionListener fires an event.
	/// </summary>
  public class ConnectionProcessor
  {
    Socket socket;

    public ConnectionProcessor(Socket theSocket)
    {
      socket = theSocket;
    }

    public void WaitForRequests()
    {
      while (true)
      {
        if (socket.Available > 0)
          FireRequestStarted(socket);

        FireIdle();
      }
    }
  		
    public delegate void RequestStartedHandler(Socket theSocket);
    public event RequestStartedHandler OnRequestStarted;
    public void FireRequestStarted(Socket theSocket)
    {
      if (OnRequestStarted != null)
        OnRequestStarted(theSocket);
    }

    public delegate void RequestProcessedHandler(string theRequest, int theSequenceNumber, int theDuration, string theResponse);
    public event RequestProcessedHandler OnRequestProcessed;
    public void FireRequestProcessed(string theRequest, int theSequenceNumber, int theDuration, string theResponse)
    {
      if (OnRequestProcessed != null)
        OnRequestProcessed(theRequest, theSequenceNumber, theDuration, theResponse);
    }

    public delegate void IdleHandler();
    public event IdleHandler OnIdle;
    public void FireIdle()
    {
      if (OnIdle != null)
        OnIdle();
    }
  }
}
